Use netip.Addr instead of net.IP

antoninbas opened the pull request:
Using the net/netip package instead of the net package can help reduce
the memory footprint of the library and help reduce the number of
heap allocations.

This is a breaking change for consumers for the libray as exported types
are updated to use fields of type netip.Addr instead of net.IP.

We also include additional benchmarks to better understand the impact
of this change.

Fixes #35

antoninbas commented:
Before the change:
```
$ make bench-integration 
sudo modprobe -a nf_nat nf_conntrack xt_conntrack xt_MASQUERADE nf_conntrack_netlink
go test -bench=. -tags=integration -exec sudo ./...
goos: linux
goarch: amd64
pkg: github.com/ti-mo/conntrack
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkExpectUnmarshal-4            	156274430	         7.752 ns/op	       0 B/op	       0 allocs/op
BenchmarkCreateDeleteFlow-4           	   34616	     34257 ns/op	   20368 B/op	     180 allocs/op
BenchmarkFlowUnmarshal-4              	  242976	      4404 ns/op	    2952 B/op	      99 allocs/op
BenchmarkStatusUnmarshalAttribute-4   	23452940	        49.31 ns/op
BenchmarkIPTupleUnmarshal/ipv4-4      	 1959086	       619.5 ns/op	      48 B/op	       6 allocs/op
BenchmarkIPTupleUnmarshal/ipv6-4      	 1991146	       595.6 ns/op	      64 B/op	       4 allocs/op
PASS
ok  	github.com/ti-mo/conntrack	10.908s

$ go test -v -run=TestFlowsMemoryUsage
=== RUN   TestFlowsMemoryUsage
    flow_test.go:560: Num flows: 100000
    flow_test.go:561: Live objects: 1100511
    flow_test.go:562: Bytes of allocated heap objects: 67448680
    flow_test.go:563: Avg bytes per flow: 674
--- PASS: TestFlowsMemoryUsage (1.05s)
PASS
ok  	github.com/ti-mo/conntrack	1.064s
```

After the change:
```
$ make bench-integration 
sudo modprobe -a nf_nat nf_conntrack xt_conntrack xt_MASQUERADE nf_conntrack_netlink
go test -bench=. -tags=integration -exec sudo ./...
goos: linux
goarch: amd64
pkg: github.com/ti-mo/conntrack
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkExpectUnmarshal-4            	155513721	         7.733 ns/op	       0 B/op	       0 allocs/op
BenchmarkCreateDeleteFlow-4           	   35670	     32901 ns/op	   20400 B/op	     188 allocs/op
BenchmarkFlowUnmarshal-4              	  262989	      4239 ns/op	    2856 B/op	      93 allocs/op
BenchmarkStatusUnmarshalAttribute-4   	23563472	        49.02 ns/op
BenchmarkIPTupleUnmarshal/ipv4-4      	 2133122	       550.4 ns/op	      16 B/op	       4 allocs/op
BenchmarkIPTupleUnmarshal/ipv6-4      	 2001910	       598.8 ns/op	      64 B/op	       4 allocs/op
PASS
ok  	github.com/ti-mo/conntrack	10.770s

$ go test -v -run=TestFlowsMemoryUsage
=== RUN   TestFlowsMemoryUsage
    flow_test.go:558: Num flows: 100000
    flow_test.go:559: Live objects: 500503
    flow_test.go:560: Bytes of allocated heap objects: 57847064
    flow_test.go:561: Avg bytes per flow: 578
--- PASS: TestFlowsMemoryUsage (0.97s)
PASS
ok  	github.com/ti-mo/conntrack	0.984s
```
 
Some observations:
* memory usage for a `Flow` object goes down from 674B to 578B
* `BenchmarkCreateDeleteFlow` shows slightly increased memory usage. This is because the benchmark keeps marshaling the same flow and the marshaling process is able to reuse the `net.IP` slices from the `Flow` object, without having to allocate new ones. With `netip.Addr`, the marshaling / serialization requires creating byte slices. However, in practice: 1) actual code is unlikely to marshal the _same_ flow over and over again, 2) I believe that most users of the library would be interested in reading conntrack entries, hence unmarshaling flows.

ti-mo commented on flow_test.go:
```diff
@@ -520,3 +521,44 @@ func BenchmarkFlowUnmarshal(b *testing.B) {
+// number of Flow objects. We don't use a Benchmark for this, because we want to
+// know memory usage *after* garbage collection has run (i.e., we want to
+// exclude allocations for temporary objects when unmarshaling).
+func TestFlowsMemoryUsage(t *testing.T) {
```
I don't think this test adds much value, could you remove this commit?

Benchmarks do provide the highest signal to noise ratio when it comes to performance and allocations. Note that a single `runtime.GC()` is not guaranteed to free all unreferenced objects. Objects with finalizers first need to have these run, and are punted to a future cycle. This process is not deterministic, and will depend on the order the `testing` package decides to execute the tests in. Only consider this a rough approximation at best.

antoninbas replied:
Yes I don't mind removing this. Benchmarks are not too great at indicating how much memory a long-running program may be consuming.

ti-mo commented on tuple.go:
```diff
@@ -87,42 +88,38 @@ func (t Tuple) marshal(at uint16) (netfilter.Attribute, error) {
 
 // An IPTuple encodes a source and destination address.
-// Both of its members are of type net.IP.
+// Both of its members are of type netip.Addr.
```
Bit of a redundant comment IMO, feel free to drop it.

ti-mo commented on flow_test.go:
```diff
@@ -514,9 +514,9 @@ func BenchmarkFlowUnmarshal(b *testing.B) {
 	for n := 0; n < b.N; n++ {
 		// Make a new copy of the AD to avoid reinstantiation.
-		iad := ad
+		iad := *ad
```
Woops. This would be a good case for `netlink.AttributeDecoder.Reset()`. Would you be willing to contribute that to Matt's library?

antoninbas replied:
I personally do not think an `AttributeDecoder` should be reused at all, which means I do not think `Reset` is a valuable addition to the public API of `AttributeDecoder`. It only makes sense here in the context of a benchmark, taking into account the fact that using `StartTimer` / `StopTimer` would not be ideal here and would make the benchmark slower.

ti-mo commented on utils_test.go:
```diff
@@ -0,0 +1,16 @@
+// cmp.Diff with the desired options. In our case, we need cmpopts.EquateComparable(netip.Addr{}) to
+// make sure that structs which contain netip.Addr fields can be compared without a panic.
+// See https://github.com/google/go-cmp/issues/339
+func Diff(x, y interface{}) string {
```
Honestly, not sure why I was using `cmp.Diff` to begin with. I think at the time assert.Equals' diff output was hard to understand, but nowadays it's just fine. Let's replace all uses of this with simple `assert.Equals` and remove our direct dependency on `go-cmp`.

antoninbas replied:
I agree. It wasn't "possible" when using `net.IP` as the semantics of `reflect.DeepEqual` for that type are different from the semantics of https://pkg.go.dev/net#IP.Equal (which is used by `cmp.Diff`).

ti-mo commented on tuple.go:
```diff
@@ -87,42 +88,38 @@ func (t Tuple) marshal(at uint16) (netfilter.Attribute, error) {
 		}
 
 		switch ipTupleType(ad.Type()) {
 		case ctaIPv4Src:
```
Fold these cases like
```suggestion
		case ctaIPv4Src, ctaIPv6Src:
```

ti-mo commented on tuple.go:
```diff
@@ -133,38 +130,29 @@ func (ipt *IPTuple) unmarshal(ad *netlink.AttributeDecoder) error {
-	if src, dest := ipt.SourceAddress.To4(), ipt.DestinationAddress.To4(); src != nil && dest != nil {
-		nfa.Children[0] = netfilter.Attribute{Type: uint16(ctaIPv4Src), Data: src}
-		nfa.Children[1] = netfilter.Attribute{Type: uint16(ctaIPv4Dst), Data: dest}
+	if ipt.SourceAddress.Is4() && ipt.DestinationAddress.Is4() {
```
This would be better expressed as a switch statement now we can use a sane API to detect the address family.

ti-mo commented on tuple_test.go:
```diff
@@ -26,6 +25,48 @@ var (
 		Nested: true, Children: []netfilter.Attribute{attrDefault}}
 )
 
+var (
```
Nit: make part of the existing `var` block?

ti-mo commented on flow_test.go:
```diff
@@ -514,9 +514,9 @@ func BenchmarkFlowUnmarshal(b *testing.B) {
 
```
Looks like we're also missing a ResetTimer() here.

ti-mo reviewed (changes requested):
Thank you! Left a few nits. Would like to get this merged so we can make a long-overdue release this week. :slightly_smiling_face: 

Note that I also made some changes today, please give this a rebase.

antoninbas commented:
@ti-mo thanks for the quick review. I will address your comments today.

Note that I think it may be a good idea to have one minor release (v0.5.0) before merging this change, and one minor release (v0.6.0) just after merging this change. v0.5.0 would be the last release to support `net.IP`, and it gives consumers of this library the opportunity to consume all the latest fixes / improvements that have been merged into the master branch, in a non-disruptive way. From v0.6.0 onwards, the API is changed (`net.IP` -> `netip.Addr`) and consumers have to update their code. The v0.5.0 and v0.6.0 release notes in Github should include the appropriate disclaimers. What do you think?

ti-mo commented:
Yep, was considering that as well. I'll bump the Go version and do a few more dependency updates, then release 0.5.0.

antoninbas commented:
@ti-mo I have addressed your comments, save for the one about `netlink.AttributeDecoder.Reset()`. I believe `AttributeDecoder` should not be re-used as a rule, and we make an exception for the benchmark. Even if you feel strongly about it, it seems orthogonal to this PR, and something that can be taken care of as a follow-up, especially since it involves another library.

ti-mo commented:
I gave this a rebase and a `go mod tidy` as go-cmp was removed. I've just released 0.4.1 and I'll release 0.5.0 when this PR is merged.

ti-mo reviewed (approved):
Thanks for taking this on. Looks great!
